dist/gaia-header.js

```diff
@@ -277,124 +456,519 @@ module.exports = Component.register('gaia-header', {
+
+    var titles = this.els.titles;
+    var space = this.getTitleSpace();
+    var styles = [].map.call(titles, (el) => this.getTitleStyle(el, space));
+
+    // Update the title styles using the latest
+    // styles. This function can be called many
+    // times but will only run once in each tick.
+    return this.setTitleStylesSoon(styles);
+  },
+
+  /**
+   * Microtask debounced `runFontFit`
    *
-   * TODO: We really need an official API for this.
+   * @private
+   */
+  runFontFitSoon: function() {
+    debug('run font-fit soon');
+    if (this.pending.runFontFitSoon) { return; }
+    this.pending.runFontFitSoon = this.nextTick(() => {
+      delete this.pending.runFontFitSoon;
+      this.runFontFit();
+    });
+  },
+
+  /**
+   * Get the style properties required
+   * to fit and position the title text.
    *
+   * @param  {HTMLH1Element} el
+   * @param  {Number} space
+   * @return {Object} {fontSize, marginStart, overflowing, id}
    * @private
    */
-  rerunFontFit: function() {
-    for (var i = 0; i < this.els.headings.length; i++) {
-      fontFit.reformatHeading(this.els.headings[i]);
+  getTitleStyle: function(el, space) {
+    debug('get el style', el, space);
+    var text = el.textContent;
+    var styleId = text + space;
+
+    // Bail when there's no text (or just whitespace)
+    if (!text || !text.trim()) { return false; }
+
+    // If neither the text or the titleSpace
+    // changed, there's no reason to continue.
+    if (getStyleId(el) === styleId) { return false; }
+
+    var marginStart = this.getTitleMarginStart();
+    var textSpace = space - Math.abs(marginStart);
+    var fontFitResult = this.fontFit(text, textSpace);
+    var overflowing = fontFitResult.textWidth > textSpace;
+
+    // If the text is overflowing in the
+    // centered title, we remove centering
+    // to free up space, rerun fontFit to
+    // get a fontSize which fits this space.
+    if (overflowing) {
+      fontFitResult = this.fontFit(text, space);
+      marginStart = 0;
     }
+
+    return {
+      fontSize: fontFitResult.fontSize,
+      marginStart: marginStart,
+      overflowing: overflowing,
+      id: styleId
+    };
   },
 
   /**
-   * Triggers the 'action' button
-   * (used in testing).
+   * Set's styles on the title elements.
    *
-   * @public
+   * If there is already an unresolved Promise
+   * we return it instead of scheduling another.
+   * This means that the function can be hammered
+   * in the same sync-turn and will only run
+   * once (like a debounce).
+   *
+   * @param {Array} styles
+   * @return {Promise}
    */
-  triggerAction: function() {
-    if (this.isSupportedAction(this.getAttribute('action'))) {
-      this.els.actionButton.click();
-    }
+  setTitleStylesSoon: function(styles) {
+    debug('set title styles soon', styles);
+    var key = 'setStyleTitlesSoon';
+
+     // Always update styles
+    this._titleStyles = styles;
```

Comment:
The goal was to avoid having an instance variable like `this._titleStyles` at all. With the proposal we have one less state. But as you said we call nextTick every time.

In the end, up to you, I'm only suggesting an alternate possibility which I think is marginally cleaner, but I see no harm with your code.